pkg/api/v1beta1/types.go

```diff
@@ -1195,3 +1195,92 @@ type SecretList struct {
 
 	Items []Secret `json:"items" description:"items is a list of secret objects"`
 }
+
+// AutoScaler monitors other resources that are resizeable and adjusts them according to configuration.
+type AutoScaler struct {
+	TypeMeta `json:",inline"`
+
+	// Spec defines the auto-scaler targets and thresholds.
+	Spec AutoScalerSpec `json:"spec,omitempty"`
+
+	// Status defines the actions the auto-scaler has taken.
+	Status AutoScalerStatus `json:"status,omitempty"`
+}
+
+// AutoScalerStatus provides the status of an auto-scaler.
+type AutoScalerStatus struct {
+	// TODO: open for discussion on what meaningful information can be reported in the status
+	// The status may return the replica count here but we may want more information
+	// such as if the count reflects a threshold being passed.
+}
+
+// AutoScalerSpec defines the auto-scaler targets and thresholds.
+type AutoScalerSpec struct {
+	// Thresholds holds a collection of AutoScaleThresholds that drive the auto-scaler.
+	Thresholds []AutoScaleThreshold `json:"thresholds,omitempty"`
+
+	// Enabled turns auto scaling on or off.
+	Enabled bool `json:"enabled,omitempty"`
+
+	// MaxAutoScaleCount defines the max replicas that the auto-scaler can use.  This value must be
+	// >= MinAutoScaleCount.
+	MaxAutoScaleCount int `json:"maxAutoScaleCount,omitempty"`
+
+	// MinAutoScaleCount defines the minimum number replicas that the auto-scaler can reduce to,
+	// 0 means that the application is allowed to idle.
+	MinAutoScaleCount int `json:"minAutoScaleCount,omitempty"`
+
+	// TargetSelector provides the resizeable target(s).  Right now this is a ReplicationController
+	// in the future it could be a job or any resource that implements resize.  If multiple targets
+	// are resolved by the selector the auto-scaler will resize the largest one.
+	TargetSelector map[string]string `json:"targetSelector,omitempty"`
```

Comment:
@smarterclayton 
so....when using labels.Set in an api type I end up with the circular dependency of

1. api imports labels
1. labels imports validations
1. validations imports api.errors
1. errors imports api

Breaking up the label types into two packages, one with `Lables`, `Set`, and `Selector` and another with all the current implementations (selectors, validations, etc) has yielded a pretty large refactoring effort considering all the places that reference selectors.

Captain, permission to abort for this PR or shall I carry on?

Just to rule out any PEBKAC, this is how I chose to refactor to try and minimize how many files were touched.  I left everything that existed in the labels package mostly untouched.  I then broke out the types in labels.go and selector.go into their own package and added some utility methods to labels.go to easily make a Labels struct without importing the types directly.  That still left me with at least 75 files to touch.  https://github.com/pweil-/kubernetes/tree/labels-as-apitype/pkg/labels